### PR TITLE
ICU-21581 integrate CLDR 40a0 to ICU trunk

### PR DESCRIPTION
**Pull request by pedberg-icu**
##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21581
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

This integrates the CLDR data we are calling release-40-a0 (for alpha 0)
- In the main CLDR repo there is no tag for this; it corresponds to CLDR master up through the following...
```
	commit cf9b65c27a6e3d710bf367946bcde5349fd25740
	Author: Mark Davis <mark@macchiato.com>
	Date:   Mon Aug 16 17:36:42 2021 -0700
		CLDR-14982 fix plural samples (#1439)
```
- ...plus the following single commit.
```
	commit 675b8784e43c780ad4750ac3d2fa5f82d87d5c6d
	Author: Peter Edberg <42151464+pedberg-icu@users.noreply.github.com>
	Date:   Wed Aug 18 10:56:27 2021 -0700
		CLDR-14976 de: compact decimals, revert formats for 1000..100000 to use 0, not Tsd. (#1449)
```
- In the cldr-staging repo there is a tag for this: "release-40-a0"

In addition to the CLDR data changes and corresponding ICU test changes, there were/are changes needed in ICU code:
- I made changes in ICU4C PluralRules::getSamples and ICU4J FixedDecimal.getShiftedValue to take the FixedDecimal exponent into account when generating sample values.
- I filed https://unicode-org.atlassian.net/browse/ICU-21714 which may end up as changes to text code or changes in PluralRules somewhere, and filed some logKnownIssue test skips using it.

### CODE REVIEW

File: icu4c/source/data/brkitr/root.txt
Comment (sffc) [inline icu4c/source/data/brkitr/root.txt:28]: Why did you delete these lines? I think you should not delete them.

File: icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd
Comment (sffc) [inline icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd:1]: Why did you add these dtd files?

File: icu4c/source/test/intltest/plurults.cpp
Comment (sffc) [inline icu4c/source/test/intltest/plurults.cpp:411]: Remove comment?

File: icu4c/source/test/intltest/plurults.cpp
Comment (sffc) [inline icu4c/source/test/intltest/plurults.cpp:472]: Remove comment?

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/PluralRulesTest.java
Comment (sffc) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/PluralRulesTest.java:751]: Remove comment?

File: tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/BreakIteratorMapper.java
Comment (sffc) [inline tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/BreakIteratorMapper.java:47]: What does this do and why did you add it now?

File: icu4c/source/data/brkitr/root.txt
Comment (markusicu) [reply icu4c/source/data/brkitr/root.txt:28]: See the email thread “Have you ever converted the lstm data to ICU4J”

File: tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/BreakIteratorMapper.java
Comment (markusicu) [reply tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/BreakIteratorMapper.java:47]: See the email thread “Have you ever converted the lstm data to ICU4J” -- Peter went partway towards adding the lstm lines via the file generator, until we said that we don't want to add this data by default. I guess this is ok to keep for later.

**Review by macchiati — APPROVED** [review]

File: icu4c/source/data/brkitr/root.txt
Comment (pedberg-icu) [reply icu4c/source/data/brkitr/root.txt:28]: I think Shane was not on that thread. These lines are not supposed to be added manually to data/brkitr/root.txt; they are only supposed be added via additions to data/xml/brkitr/root.xml (and if not there, the integration process will delete the lines in root.txt). In order to start the process of adding the support in data/xml/brkitr/root.xml, I had to make some dtd additions and changes to tool/cldrtoicu/mapper/BreakIteratorMapper.java. But then we decided to not automatically include the lstm data; consequently I commented out the relevant additions in data/xml/brkitr/root.xml (if I do not do that, the creation of ICU4J jars fails with an unresolved dependency).

File: icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd
Comment (pedberg-icu) [reply icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd:1]: See above comment

File: icu4c/source/test/intltest/plurults.cpp
Comment (pedberg-icu) [reply icu4c/source/test/intltest/plurults.cpp:411]: Since the ICU ticket is not resolved, I was worried that the problem may depend on values that happen to be in samples. I wanted to leave this there in case we need it again.

File: icu4c/source/test/intltest/plurults.cpp
Comment (pedberg-icu) [reply icu4c/source/test/intltest/plurults.cpp:472]: See above. Since the ICU ticket is not resolved, I was worried that the problem may depend on values that happen to be in samples. I wanted to leave this there in case we need it again.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/PluralRulesTest.java
Comment (pedberg-icu) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/PluralRulesTest.java:751]: See above.

File: tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/BreakIteratorMapper.java
Comment (pedberg-icu) [reply tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/BreakIteratorMapper.java:47]: See my first comment. This was about adding support for the *right* way to include the lstm data when we decide to do that.

File: icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd
Comment (markusicu) [reply icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd:1]: There is "cldr-40" in the path. We already have these in a "cldr-35.1" path. Maybe we need these each time there is a certain change to the DTD?? Should the new ones *replace* the old ones?

Anyway, files like icu4c/source/data/xml/collation/root.xml refer to the new DTDs now. I didn't check if there are still files referring to the old ones.

File: icu4j/main/classes/core/src/com/ibm/icu/text/PluralRules.java
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/text/PluralRules.java:1081]: *if* you amend anyway... s/taxe/take/ -- otherwise not important

File: icu4c/source/data/brkitr/root.txt
Comment (sffc) [reply icu4c/source/data/brkitr/root.txt:28]: Thanks Markus for forwarding the thread.

***It seems like removing these lines breaks the LSTM.***  Frank added these lines to root.txt as an optional dependency.  If the tooling doesn't like these lines, then we should open a ticket to fix the tooling, and consider these lines as a temporary patch until the tooling is fixed.  Evidence: on mainline, the TestLSTMThai test takes 180 ms to run ([source](https://github.com/unicode-org/icu/runs/3366503463)):

```
2021-08-18T23:30:15.1191864Z          TestLSTMThai {
2021-08-18T23:30:15.2987089Z          } OK:   TestLSTMThai  (180ms) 
```

However, on this PR, the test takes less than 1 ms, suggesting that the test gets skipped ([source](https://github.com/unicode-org/icu/pull/1802/checks?check_run_id=3366784491)):

```
2021-08-19T00:18:34.7485907Z          TestLSTMThai {
2021-08-19T00:18:34.7486498Z          } OK:   TestLSTMThai 
```

Here is the "skip test" line taht gets run unless the data is present in the bundle:

https://github.com/unicode-org/icu/blob/c26aebe802f1a87483f70ad7557fa0ce52061eb9/icu4c/source/test/intltest/rbbitst.cpp#L5261

**Review by sffc — DISMISSED** [review]

File: tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/BreakIteratorMapper.java
Comment (sffc) [reply tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/BreakIteratorMapper.java:47]: See reply above.  I think we should separate the task of adopting the "right way" of adding the LSTM data from the task of updating to CLDR 40.

**Comment by markusicu** [comment]
@sffc @FrankYFTang @pedberg
> It seems like removing these lines breaks the LSTM. Frank added these lines to root.txt as an optional dependency.

Yes. We discussed this today and decided that, since the LSTMs are not included by default, it is ok for now to remove these lines.

> If the tooling doesn't like these lines, then we should open a ticket to fix the tooling, and consider these lines as a temporary patch until the tooling is fixed.

Yes, there should be a ticket.
Manually keeping/restoring a patch each time is cumbersome for the several rounds of CLDR integration to come over the next few weeks.

> Evidence: on mainline, the TestLSTMThai test takes 180 ms to run

I don't think any of us was thinking of the non-default test configuration that adds the LSTMs for testing. So with these lines gone we lose test coverage. :-(

Still, I don't think we should hold up this PR here if this is the only outstanding issue. (Right now there are also questions about the DTD files.)

**Comment by sffc** [comment]
> Still, I don't think we should hold up this PR here if this is the only outstanding issue. (Right now there are also questions about the DTD files.)

My concrete suggestion on how to unblock this PR is to restore the four deleted lines in root.txt (now and until the root cause is fixed).

We discussed at length earlier in the cycle about how to enable and disable the LSTM. What is currently in the repo is what we agreed upon. We added a flag to the data build tool and it was reviewed by multiple committee members. It is even backed by a CI test.

If Peter proposes changing the process to enable the LSTM, then he should, at a minimum, also update the CI bot to apply the patch.

However, I still strongly prefer manually retaining the four lines.

**Comment by sffc** [comment]
> since the LSTMs are not included by default, it is ok for now to remove these lines.

This is not entirely correct. Including the LSTM only requires a custom filters.json file. It does NOT require patching the source. It was an intentional part of the design that someone could check out ICU 70 and enable the LSTM via a build configuration without patching the sources.

**Comment by markusicu** [comment]
> We discussed at length earlier in the cycle about how to enable and disable the LSTM. What is currently in the repo is what we agreed upon. We added a flag to the data build tool and it was reviewed by multiple committee members. It is even backed by a CI test.

Yes, you are right. However, we all failed to remember or realize that the CLDR-to-ICU converter generates this root.txt file, so manually changing it turned out to be brittle and error-prone.

Peter is pressed for time right now, and no one else is prepared to shepherd the CLDR-to-ICU integration. Pragmatically although not ideally, I think we should let him drop these lines here, and let Peter+Frank revisit it soon (before rc).

File: icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd
Comment (markusicu) [reply icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd:1]: @pedberg could you please take a look at this question, too? This might be the one other place where we might want changes. (And why are these DTD files in a versioned path in the ICU repo anyway?)

**Comment by sffc** [comment]
In the interest of not standing in the way of the complicated integration process, I'll unblock this PR, but I want to see a blocks-release ticket filed to follow up with the LSTM before RC.

File: icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd
Comment (pedberg-icu) [reply icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd:1]: I do not know why they are in a versioned path. But there were dtd updates for 40 that were needed for the data/xml files.

Comment (pedberg-icu) [reply icu4c/source/data/dtd/cldr-40/common/dtd/ldmlICU.dtd:1]: Any chnage we want here I would like to do in the next PR for alpha1 so I can merged this one before I start alpha1 integration.

File: icu4c/source/data/brkitr/root.txt
Comment (pedberg-icu) [reply icu4c/source/data/brkitr/root.txt:28]: I can re-add the lines but I will do that as part of the alpha1 integration, which I am starting now and should have done within a day. I want to merge this PR before I start that integration so I do not want to make those changes in this PR. I am running out of time.

**Comment by pedberg-icu** [comment]
> In the interest of not standing in the way of the complicated integration process, I'll unblock this PR, but I want to see a blocks-release ticket filed to follow up with the LSTM before RC.

As I have said, I am immediately going to do another integration, so I will add these lines as part of that. They do not need to be in this PR.

**Comment by pedberg-icu** [comment]
I will also fix the 6 exhaustive test failures in the alpha1 integration.
```
2021-08-19T02:55:40.7332695Z !!    getDayNames(mai, 4) at index 2 returned characters not in the exemplar characters: 0943.
2021-08-19T02:55:40.7333664Z !!    getDayNames(mai_IN, 4) at index 2 returned characters not in the exemplar characters: 0943.
2021-08-19T02:55:40.7551684Z !!    getDayNames(sd_Deva, 3) at index 3 returned characters not in the exemplar characters: 0952.
2021-08-19T02:55:40.7552509Z !!    getMonthNames(sd_Deva, 9) at index 1 returned characters not in the exemplar characters: 0911.
2021-08-19T02:55:40.7553307Z !!    getDayNames(sd_Deva_IN, 3) at index 3 returned characters not in the exemplar characters: 0952.
2021-08-19T02:55:40.7553943Z !!    getMonthNames(sd_Deva_IN, 9) at index 1 returned characters not in the exemplar characters: 0911.
```

**Pull request merged**
